### PR TITLE
Fix errors for apple pay registration

### LINKED ISSUE #4511: Apple Pay: fix errors on web registrations
### Describe the bug

When a domain is registered, we saw a burst of domain registration errors.
Possible causes: 
1. `apple_pay_bad_status_code ` - meaning when `<domain>/.well-known/apple-developer-merchantid-domain-association` is accessed, we get a non 200 HTTP response
2. `apple_pay_unsupported_country` - meaning the merchant is in an unsupported country. Supported countries: https://support.stripe.com/questions/country-availability-for-apple-pay 

Context: p1657312179990539-slack-C9976E5MJ

### PR DESCRIPTION
**Pull request by dpaun1985**
Fixes #4511

#### Changes proposed in this Pull Request
- when merchant is not in the Apple Pay [supported country list](https://support.apple.com/en-us/HT207957
), or `apple-developer-merchantid-domain-association` file has different permissions than 644, the apple domain registration is not called anymore
- for the conditions above, a notice is shown in the WooCommerce Payments admin panel

Screenshot 2022-08-09 at 14 37 20

#### Testing instructions
1. Set the merchant country other than the ones listed in https://support.apple.com/en-us/HT207957 .
2. Enable Apple Pay
3. `Apple Pay: You are not in the countries and regions that support Apple Pay. You can find a full list of countries and regions at https://support.apple.com/en-us/HT207957` should be displayed in Payment methods panel
4. Rename `docker/wordpress/.well-known/apple-developer-merchantid-domain-association` file
5. `Apple Pay: Invalid domain association file permissions. /var/www/html/.well-known needs to have 755 permissions and /var/www/html/.well-known/apple-developer-merchantid-domain-association 644. Set the proper permissions and disable/enable Apple Pay.`  should be displayed in Payment methods panel

*

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

File: changelog/fix-4511-apple-pay-registration [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: fix
+
+Fix errors for apple pay registration
```
Comment (kalessil) [inline changelog/fix-4511-apple-pay-registration:4]: Can we add here what causes the errors please?

File: includes/class-wc-payment-gateway-wcpay.php [hunk 2]
```diff
@@ -594,6 +596,62 @@ public function display_test_mode_notice() {
...
+	 * @return void
+	 */
+	public function display_not_supported_apple_pay() {
+		if ( 'yes' !== $this->get_option( 'payment_request' ) ) {
+			return;
+		}
+
+		$is_in_supported_countries               = in_array( $this->account->get_account_country(), WC_Payments_Utils::supported_applepay_country_codes(), true );
+		$well_known_dir                          = untrailingslashit( ABSPATH ) . '/' . WC_Payments_Apple_Pay_Registration::DOMAIN_ASSOCIATION_FILE_DIR;
+		$fullpath                                = $well_known_dir . '/' . WC_Payments_Apple_Pay_Registration::DOMAIN_ASSOCIATION_FILE_NAME;
```
Comment (kalessil) [inline includes/class-wc-payment-gateway-wcpay.php:611]: `full_path` perhaps?

File: includes/class-wc-payment-gateway-wcpay.php [hunk 3]
```diff
@@ -594,6 +596,62 @@ public function display_test_mode_notice() {
...
+		$has_domain_association_file_permissions = is_dir( $well_known_dir ) &&
```
Comment (kalessil) [inline includes/class-wc-payment-gateway-wcpay.php:612]: Would it be possible to use `is_writable` and `is_readable` here?

File: includes/class-wc-payment-gateway-wcpay.php [hunk 4]
```diff
@@ -594,6 +596,62 @@ public function display_test_mode_notice() {
...
+
+		if ( $is_in_supported_countries && $has_domain_association_file_permissions ) {
+			return;
+		}
+		if ( ! $is_in_supported_countries ) {
+			?>
+			<div id="wcpay-applepay-error" class="notice notice-error">
+				<p>
+					<b><?php esc_html_e( 'Apple Pay: ', 'woocommerce-payments' ); ?></b>
+					<?php esc_html_e( 'You are not in the countries and regions that support Apple Pay. You can find a full list of countries and regions at', 'woocommerce-payments' ); ?>
```
Comment (kalessil) [inline includes/class-wc-payment-gateway-wcpay.php:625]: Would the following be a good alternative `Your account is registered in a country or region not supported by Apply Pay.` where `not supported` is the link.

File: includes/class-wc-payment-gateway-wcpay.php [hunk 5]
```diff
@@ -594,6 +596,62 @@ public function display_test_mode_notice() {
...
+			</div>
+			<?php
+		}
+
+		if ( ! $has_domain_association_file_permissions ) {
+			?>
+		<div id="wcpay-applepay-error" class="notice notice-error">
+			<p>
+				<b><?php esc_html_e( 'Apple Pay: ', 'woocommerce-payments' ); ?></b>
+				<?php esc_html_e( 'Invalid domain association file permissions.', 'woocommerce-payments' ); ?>
```
Comment (kalessil) [inline includes/class-wc-payment-gateway-wcpay.php:637]: `Domain association file is not readable, please check its access permissions.`

File: includes/class-wc-payment-gateway-wcpay.php [hunk 6]
```diff
@@ -594,6 +596,62 @@ public function display_test_mode_notice() {
...
+				<?php
+				echo esc_html(
+					sprintf(
+						/* translators: 1: well known dir, 2: full path */
+						__( '%1$s needs to have 755 permissions and %2$s 644.', 'woocommerce-payments' ),
```
Comment (kalessil) [inline includes/class-wc-payment-gateway-wcpay.php:642]: @`The recommended access permissions for %1$s is 0755 and for %2$s 0644.`

File: includes/class-wc-payment-gateway-wcpay.php [hunk 7]
```diff
@@ -594,6 +596,62 @@ public function display_test_mode_notice() {
...
+						$well_known_dir,
+						$fullpath
+					)
+				);
+				?>
+				<?php esc_html_e( 'Set the proper permissions and disable/enable Apple Pay. ', 'woocommerce-payments' ); ?>
```
Comment (kalessil) [inline includes/class-wc-payment-gateway-wcpay.php:648]: `Please re-enable Apple Pay after applying the new access permissions.`

File: includes/class-wc-payments-apple-pay-registration.php [hunk 8]
```diff
@@ -254,11 +254,33 @@ private function get_gateway_mode_string() {
...
 
+	/**
+	 * Check if merchant can register domain with Apple Pay.
+	 *
+	 * @return bool
+	 */
+	private function can_merchant_register_to_applepay(): bool {
+		$is_in_supported_countries               = in_array( $this->account->get_account_country(), WC_Payments_Utils::supported_applepay_country_codes(), true );
+		$well_known_dir                          = untrailingslashit( ABSPATH ) . '/' . self::DOMAIN_ASSOCIATION_FILE_DIR;
+		$fullpath                                = $well_known_dir . '/' . self::DOMAIN_ASSOCIATION_FILE_NAME;
```
Comment (kalessil) [inline includes/class-wc-payments-apple-pay-registration.php:265]: `full_path`

File: includes/class-wc-payments-apple-pay-registration.php [hunk 9]
```diff
@@ -254,11 +254,33 @@ private function get_gateway_mode_string() {
...
+		$has_domain_association_file_permissions = is_dir( $well_known_dir ) &&
```
Comment (kalessil) [inline includes/class-wc-payments-apple-pay-registration.php:266]: Would it be possible to use `is_writable` and `is_readable` here?

File: includes/class-wc-payments-apple-pay-registration.php [hunk 10]
```diff
@@ -254,11 +254,33 @@ private function get_gateway_mode_string() {
 		return 'live';
 	}
 
+	/**
+	 * Check if merchant can register domain with Apple Pay.
+	 *
+	 * @return bool
+	 */
+	private function can_merchant_register_to_applepay(): bool {
```
Comment (kalessil) [inline includes/class-wc-payments-apple-pay-registration.php:262]: `can_merchant_register_domain_with_applepay`

**Review by kalessil — CHANGES_REQUESTED** [review]
Looks good overall, left some suggestions related to readability.

File: includes/class-wc-payment-gateway-wcpay.php [hunk 11]
```diff
@@ -594,6 +596,67 @@ public function display_test_mode_notice() {
...
+		$well_known_dir                          = untrailingslashit( ABSPATH ) . '/' . WC_Payments_Apple_Pay_Registration::DOMAIN_ASSOCIATION_FILE_DIR;
+		$full_path                               = $well_known_dir . '/' . WC_Payments_Apple_Pay_Registration::DOMAIN_ASSOCIATION_FILE_NAME;
+		$has_domain_association_file_permissions = is_dir( $well_known_dir ) &&
+			is_writable( $well_known_dir ) &&
+			file_exists( $full_path ) &&
+			is_readable( $full_path );
+
+		if ( $is_in_supported_countries && $has_domain_association_file_permissions ) {
+			return;
+		}
```
Comment (mordeth) [inline includes/class-wc-payment-gateway-wcpay.php:619]: We are missing an empty line between if statements.

**Review by mordeth — APPROVED** [review]
LGTM!

File: includes/class-wc-payments-apple-pay-registration.php [hunk 12]
```diff
@@ -254,11 +254,33 @@ private function get_gateway_mode_string() {
 		return 'live';
 	}
 
+	/**
+	 * Check if merchant can register domain with Apple Pay.
+	 *
+	 * @return bool
+	 */
+	private function can_merchant_register_domain_with_applepay(): bool {
```
Comment (dmvrtx) [inline includes/class-wc-payments-apple-pay-registration.php:262]: May I suggest to move this code into the utilities file below? It is repeated here and above, and it'd be better to be in one place.

**Review by mgascam — APPROVED** [review]
LGTM! 
✅ Code looks good
✅ Followed testing instructions and works as expected

**Pull request merged**